engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -84,54 +102,71 @@ class EngineJobExecutionActor(jobKey: BackendJobDescriptorKey,
+      recordHashError(t)
+      runJob(jobDescriptor, bjeaProps)
+    case Event(CacheMiss, JobDescriptorData(jobDescriptor, bjeaProps)) =>
+      writeToMetadata(Map(callCachingReadResultMetadataKey -> "Cache Miss"))
       log.info(s"Cache miss for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}")
       runJob(jobDescriptor, bjeaProps)
-    case Event(CacheHit(cacheResultId), EJEAJobDescriptorData(Some(jobDescriptor), _)) =>
+    case Event(CacheHit(cacheResultId), JobDescriptorData(jobDescriptor, _)) =>
+      writeToMetadata(Map(callCachingReadResultMetadataKey -> s"Cache Hit (from result ID $cacheResultId)"))
       log.info(s"Cache hit for job ${jobDescriptor.key.call.fullyQualifiedName}, index ${jobDescriptor.key.index}! Copying cache result $cacheResultId")
       lookupCachedResult(jobDescriptor, cacheResultId)
   }
 
   // When RunningJob, the FSM always has EJEAPartialCompletionData (which might be None, None)
   when(RunningJob) {
-    case Event(response: SucceededResponse, EJEAPartialCompletionData(None, Some(hashes))) if callCachingMode.writeToCache =>
-      saveCacheResults(EJEASuccessfulCompletionDataWithHashes(response, hashes))
-    case Event(response: SucceededResponse, data @ EJEAPartialCompletionData(None, None)) if callCachingMode.writeToCache =>
-      stay using data.copy(jobResult = Option(response))
-    case Event(hashes: CallCacheHashes, data @ EJEAPartialCompletionData(Some(response: SucceededResponse), None)) =>
-      saveCacheResults(EJEASuccessfulCompletionDataWithHashes(response, hashes))
-    case Event(hashes: CallCacheHashes, data @ EJEAPartialCompletionData(None, None)) =>
-      stay using data.copy(hashes = Option(hashes))
-    case Event(response: BackendJobExecutionResponse, data: EJEAPartialCompletionData) =>
+    case Event(hashes: CallCacheHashes, PartialCompletionDataWithSucceededResponse(response: SucceededResponse)) =>
+      saveCacheResults(CacheWriteOnCompletionData(response, hashes))
+    case Event(hashes: CallCacheHashes, EmptyPartialCompletionData) =>
+      log.info(s"Got hashes for ${jobKey.toString(workflowId)}, awaiting job completion")
+      stay using PartialCompletionDataWithHashes(Success(hashes))
+
+    case Event(HashError(t), PartialCompletionDataWithSucceededResponse(response)) =>
+      recordHashError(t)
+      saveJobCompletionToJobStore(response)
+    case Event(HashError(t), EmptyPartialCompletionData) =>
+      recordHashError(t)
+      stay using PartialCompletionDataWithHashes(Failure(t))
+
+    case Event(response: SucceededResponse, PartialCompletionDataWithHashes(Success(hashes))) if effectiveCallCachingMode.writeToCache =>
+      saveCacheResults(CacheWriteOnCompletionData(response, hashes))
+    case Event(response: SucceededResponse, EmptyPartialCompletionData) if effectiveCallCachingMode.writeToCache =>
+      log.info(s"Got job result for ${jobKey.toString(workflowId)}, awaiting hashes")
```

Comment:
Be consistent with your logging style (this form and the slightly zippier format down below a bit)